CRM/Contact/BAO/Contact/Permission.php

```diff
@@ -335,6 +439,104 @@ public static function relationship($selectedContactID, $contactID = NULL) {
+    $result_set = array();
+
+    // no processing empty lists (avoid SQL errors as well)
+    if (empty($contact_ids)) {
+      return array();
+    }
+
+    // get the currently logged in user
+    $config    = CRM_Core_Config::singleton();
+    $session   = CRM_Core_Session::singleton();
+    $contactID = (int) $session->get('userID');
+    if (empty($contactID)) {
+      return array();
+    }
+
+    // compile a list of queries (later to UNION)
+    $queries = array();
+    $contact_id_list = implode(',', $contact_ids);
+
+    // add a select statement for each direection
+    $directions = array(array('from' => 'a', 'to' => 'b'), array('from' => 'b', 'to' => 'a'));
+
+    // NORMAL/SINGLE DEGREE RELATIONSHIPS
+    foreach ($directions as $direction) {
+      $user_id_column    = "contact_id_{$direction['from']}";
+      $contact_id_column = "contact_id_{$direction['to']}";
+
+      // add clause for deleted contacts, if the user doesn't have the permission to access them
+      $LEFT_JOIN_DELETED = $AND_CAN_ACCESS_DELETED = '';
+      if (!CRM_Core_Permission::check('access deleted contacts')) {
+        $LEFT_JOIN_DELETED       = "LEFT JOIN civicrm_contact ON civicrm_contact.id = {$contact_id_column} ";
+        $AND_CAN_ACCESS_DELETED  = "AND civicrm_contact.is_deleted = 0";
+      }
+
+      $queries[] = "
+SELECT DISTINCT(civicrm_relationship.{$contact_id_column}) AS contact_id
+  FROM civicrm_relationship 
+  {$LEFT_JOIN_DELETED}
+ WHERE civicrm_relationship.{$user_id_column} = {$contactID}
+   AND civicrm_relationship.{$contact_id_column} IN ({$contact_id_list})
+   AND civicrm_relationship.is_active = 1
+   AND civicrm_relationship.is_permission_{$direction['from']}_{$direction['to']} = 1
+   $AND_CAN_ACCESS_DELETED";
+    }
+
+    if ($config->secondDegRelPermissions) {
+      // ADD SECOND DEGREE RELATIONSHIPS
+      if ($config->secondDegRelPermissions) {
+        foreach ($directions as $first_direction) {
+          foreach ($directions as $second_direction) {
+            // add clause for deleted contacts, if the user doesn't have the permission to access them
+            $LEFT_JOIN_DELETED = $AND_CAN_ACCESS_DELETED = '';
+            if (!CRM_Core_Permission::check('access deleted contacts')) {
+              $LEFT_JOIN_DELETED       = "LEFT JOIN civicrm_contact first_degree_contact  ON first_degree_contact.id  = second_degree_relationship.contact_id_{$second_direction['from']}\n";
+              $LEFT_JOIN_DELETED      .= "LEFT JOIN civicrm_contact second_degree_contact ON second_degree_contact.id = second_degree_relationship.contact_id_{$second_direction['to']} ";
+              $AND_CAN_ACCESS_DELETED  = "AND first_degree_contact.is_deleted = 0\n";
+              $AND_CAN_ACCESS_DELETED .= "AND second_degree_contact.is_deleted = 0 ";
+            }
+
+            $queries[] = "
+  SELECT DISTINCT(second_degree_relationship.contact_id_{$second_direction['to']}) AS contact_id
```

Comment:
removed ``SELECT DISTINCT`` and used ``UNION DISTINCT``, although that was the default for ``UNION`` anyway